posts/Yoann/2019-11-08-hibernate-search-quarkus.adoc

```diff
@@ -0,0 +1,1428 @@
+----
+
+Since Hibernate Search updates the index, "lapin" will no longer match:
+
+[source]
+----
+$ curl -X GET 'http://localhost:8080/client/search/?terms=lapin'
+
+[
+]
+----
+
+\... but "vance" will match:
+
+[source]
+----
+$ curl -X GET 'http://localhost:8080/client/search/?terms=vance'
+
+[
+    {
+        "assignedManager": {
+            "email": "pvance@dundermifflin.net",
+            "id": 6,
+            "name": "Phyllis Vance"
+        },
+        "id": 7,
+        "name": "Stark Industries"
+    },
+    {
+        "assignedManager": {
+            "email": "pvance@dundermifflin.net",
+            "id": 6,
+            "name": "Phyllis Vance"
+        },
+        "id": 8,
+        "name": "Parker Industries"
+    }
+]
+----
+
+== Running in a container
+
+When the project was created, Quarkus added Dockerfiles to containerize the application
+either in JVM mode or as a native binary.
+
+However, in order to spare ourselves the download and installation of GraalVM,
+we will simply use a https://quarkus.io/guides/building-native-image.html#creating-a-container-with-a-multi-stage-docker-build[multi-stage Docker build]
+that will build our application in a container, then generate a container for our application.
+
+Let's add a Dockerfile at `src/main/docker/Dockerfile.multistage`:
+
+[source]
+----
+## Stage 1 : build with maven builder image with native capabilities
+FROM quay.io/quarkus/centos-quarkus-maven:19.2.1 AS build
+COPY src /usr/src/app/src
+COPY pom.xml /usr/src/app
+USER root
+RUN chown -R quarkus /usr/src/app
+USER quarkus
+RUN mvn -f /usr/src/app/pom.xml -Pnative clean package
+
+## Stage 2 : create the docker final image
+FROM registry.access.redhat.com/ubi8/ubi-minimal
+WORKDIR /work/
+COPY --from=build /usr/src/app/target/*-runner /work/application
+RUN chmod 775 /work
+EXPOSE 8080
+CMD ["./application", "-Dquarkus.http.host=0.0.0.0"]
+----
+
+Then let's build it (it will take some time, we're compiling a native binary here):
+
+[source]
+----
+$ docker build -f src/main/docker/Dockerfile.multistage -t quarkus/hsearch-quarkus .
+[... lots of logs ...]
+Successfully tagged quarkus/hsearch-quarkus:latest
+----
+
+The Docker image is now ready to be used.
+Let's start an environment if it's not already done:
+
+[source]
+----
+$ docker-compose -f environment-stack.yml -p hsearch-quarkus-env up
+----
+
+And once everything is ready, let's start our application:
+
+[source]
+----
+$ docker run --rm -it --network=host \
+        -e POSTGRES_HOST=localhost \
+        -e POSTGRES_DB=hsearch_demo \
+        -e POSTGRES_USER=hsearch_demo \
+        -e POSTGRES_PASSWORD=hsearch_demo \
+        -e ES_HOSTS=http://localhost:9200 \
+        quarkus/hsearch-quarkus
+2019-11-07 10:33:15,520 INFO  [io.quarkus] (main) hsearch-quarkus 1.0-SNAPSHOT (running on Quarkus 1.0.0.CR1) started in 0.849s. Listening on: http://0.0.0.0:8080
```

Comment:
This is with a native image? Looks weird that it doesn't start much faster. You are in prod mode right so you don't inject the data?

Reply:
I don't inject the data, but I initialize the schema. I believe that's why it's so slow. I guess I could write a few more scripts to initialize the DB and Elasticsearch, but I'm not sure it's reasonable to assume people will initialize the Elasticsearch schema by hand.

Strangely enough, the second run is a bit faster (0.5s) and the next ones much faster (0.1). Maybe it's related to Elasticsearch, I've noticed it takes a while to "warm up" and deliver decent response times.